src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1121,10 +1165,86 @@ private static ImmutableArray<StateForCase> RemoveEvaluation(ImmutableArray<Stat
             falseTestPermitsTrueOther = true;
             trueTestImpliesTrueOther = false;
             falseTestImpliesTrueOther = false;
+            precondition = Tests.True.Instance;
+            sideeffect = Tests.True.Instance;
 
             // if the tests are for unrelated things, there is no implication from one to the other
-            if (!test.Input.Equals(other.Input))
-                return;
+            BoundDagTemp s1Input = test.Input;
+            BoundDagTemp s2Input = other.Input;
+            if (!s1Input.Equals(s2Input))
+            {
+                var preconditions = ArrayBuilder<Tests>.GetInstance();
+                do
```

Comment:
I think it would be helpful to explain and have an example of why this needs to be a loop. It's not immediately obvious to me.